docs/integrations/custom_middleware.md

```diff
@@ -0,0 +1,191 @@
+# Writing your own Custom Middleware
+
+## How to write your own custom middleware for Mirascope
+
+`middleware_decorator` is a helper function to assist in helping you wrap any Mirascope call.
+We will be creating an example decorator `with_saving` that saves some metadata after a Mirascope call:
+
+### Writing the decorator
+
+```python
+from mirascope.integrations import middleware_decorator
+
+def with_saving(fn):
+    """Saves some data after a Mirascope call."""
+
+    return middleware_decorator(
+        fn,
+        custom_context_manager=custom_context_manager,
+        custom_decorator=custom_decorator,
+        handle_call_response=handle_call_response,
+        handle_call_response_async=handle_call_response_async,
+        handle_stream=handle_stream,
+        handle_stream_async=handle_stream_async,
+        handle_response_model=handle_response_model,
+        handle_response_model_async=handle_response_model_async,
+        handle_structured_stream=handle_structured_stream,
+        handle_structured_stream_async=handle_structured_stream_async,
+    )
+```
+
+We will go over each of the different functions.
+
+### `handle_call_response` and `handle_call_response_async`
+
+These functions will be called after making a mirascope call with the following signature:
+
+```python
+from typing import Callable
+
+from mirascope.core.base import BaseCallResponse
+
+
+def handle_call_response(
+    result: BaseCallResponse, fn: Callable, context_manager: ContextManagerType | None
+):
+    # handle after call here
+```
+
+The first argument will be a Mirascope `CallResponse` of the provider you are using.
```

Comment:
IIUC these won't actually show up on separate lines unless theres two newlines? Do you want these on a single line and are just writing it this way for readability in the markdown?